subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+				continue;
+			}
+
+			/* Configure new audio block */
+			block.data = data;
+			block.data_size = handle->thread.data_size;
+
+			/* Process the input audio block */
+			ret = handle->description->functions->data_process(
+				(struct audio_module_handle_private *)handle, NULL, &block);
+			if (ret) {
+				k_mem_slab_free(handle->thread.data_slab, (void **)(&data));
+
+				LOG_DBG("Data process error in module %s, ret %d", handle->name,
+					ret);
+				continue;
+			}
+
+			/* Send input audio block to next module(s) */
+			send_to_connected(handle, &block);
+		} else {
+			LOG_DBG("No process function for module %s, no input", handle->name);
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief The thread that processes inputs and outputs them out of the audio system.
+ *
+ * @param handle  The handle for this modules instance
+ *
+ * @return 0 if successful, error value
+ */
+static int module_thread_output(struct audio_module_handle *handle, void *p2, void *p3)
+{
+	int ret;
+
+	struct audio_module_message *msg_rx;
+	size_t size;
+
+	if (handle == NULL) {
+		LOG_DBG("Module task has NULL handle");
+		return -EINVAL;
+	}
+
+	/* Execute thread */
+	while (1) {
+		msg_rx = NULL;
+
+		ret = data_fifo_pointer_last_filled_get(handle->thread.msg_rx, (void **)&msg_rx,
+							&size, K_FOREVER);
+		if (ret) {
+			LOG_DBG("No new data for module %s, ret %d", handle->name, ret);
+			continue;
+		}
+
+		if (handle->description->functions->data_process != NULL) {
+			/* Process the input audio block and output from the audio system */
+			ret = handle->description->functions->data_process(
+				(struct audio_module_handle_private *)handle, &msg_rx->block, NULL);
+			if (ret) {
+				if (msg_rx->response_cb != NULL) {
+					msg_rx->response_cb((struct audio_module_handle_private *)
+								    msg_rx->tx_handle,
+							    &msg_rx->block);
+				}
+
+				LOG_DBG("Data process error in module %s, ret %d", handle->name,
+					ret);
+				continue;
+			}
+
+			if (msg_rx->response_cb != NULL) {
+				msg_rx->response_cb(
+					(struct audio_module_handle_private *)msg_rx->tx_handle,
+					&msg_rx->block);
+			}
+		} else {
+			LOG_DBG("No process function for module %s, discarding input",
+				handle->name);
+		}
+
+		data_fifo_block_free(handle->thread.msg_rx, (void **)&msg_rx);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief The thread that processes inputs and outputs the data from the module
+ *
+ * @param handle  The handle for this modules instance
+ *
+ * @return 0 if successful, error value
+ */
+static void module_thread_in_out(void *p1, void *p2, void *p3)
+{
+	int ret;
+	struct audio_module_handle *handle = (struct audio_module_handle *)p1;
+	struct audio_module_message *msg_rx;
+	struct audio_data block;
+	void *data;
+	size_t size;
+
+	LOG_DBG("Module %s task started", handle->name);
+
+	/* Execute thread */
+	while (1) {
+		data = NULL;
+
+		LOG_DBG("Module %s waiting for a message", handle->name);
```

Comment:
Would it me more descriptive to say that the module is waiting for audio data?

Reply:
Nice. This can also be printed in the "IN" module for consistency.

Reply:
Yes but should be the "OUT" module/thread,